### PR TITLE
Features/insights 3.0 system metadata

### PR DESCRIPTION
**Pull request by jeudy100**
screenshot from 2017-07-28 15-57-33

screenshot from 2017-07-28 15-52-05

@iphands @Kinlaw @jharting 
Please also look at https://github.com/RedHatInsights/dmidecode-vendor-mappings which is the npm module that is used to determine if the system is virtual or physical.

### CODE REVIEW

**Comment by Kinlaw** [comment]
Going along with jharting's comment, we should be rebasing instead of merging so that we have one linear trail of commits and so that we don't have extraneous merge commits in the git history.

**Comment by jharting** [comment]
My virtual machines running CentOS on KVM with libvirt show up as physical although they are not.

**Comment by iphands** [comment]
@jeudy100 Can you kill with :fire: the for and while loops?

**Comment by jeudy100** [comment]
@iphands i have killed with :fire: the for and while loops

**Comment by iphands** [comment]
This looks good, but it does not seem to work.
I can't get any machines to show up as physical or virtual. Also, I can't see any with networking data in the live DB.

I think we should change the order of the tabs, and also disable or even hide the Networking tab if there is no data to see.

**Comment by mattnolting** [comment]
Before this is merged, I need to complete a more thorough code review than I'm currently able to do (at the latest 8/31). It's not quite ready yet.

File: app/js/components/systemMetadata/systemMetadata.jade [hunk 1]
```diff
@@ -1,11 +1,104 @@
...
+      .col-sm-5
+        table.meta-table
+          tbody
+            tr(ng-repeat='facts in initialMetadata')
+              td
+                b(translate) {{facts.label}}
+              td(translate) {{facts.value}}
+
+#expandableDataTable.expandable-data-table(ng-if='!loading; hasMetadata()', ng-class="{expanded: expandedTabs || expandIndefinite}")
+  // Toggle button
```
Comment (mattnolting) [inline app/js/components/systemMetadata/systemMetadata.jade]: Here are the states
- User hovers over an icon in the utility bar > the text name appears, this should be handled with css. We originally discussed setting a timeout on the hover, but it feels odd and adds unnecessary complexity.
- User hovers over the expand icon. The icon should have a hover state that is somewhere between the expanded and default states.
- The list is expanded > the expand icon should be styled (with higher specificity than the hover state) to be dark/reversed text. This is when the 'expanded' class should be added to the component.

File: app/js/components/systemMetadata/systemMetadata.jade [hunk 2]
```diff
@@ -1,11 +1,104 @@
...
+  .toggle(ng-class="{expandedIndefinite: expandIndefinite}")
+    i.fa.fa-plus#iconSpin(ng-click="getData('network'); toggleTabsIndefinite(); rotateIcon()")
+
+  // Nav tabs
```
Comment (mattnolting) [inline app/js/components/systemMetadata/systemMetadata.jade]: When the user hovers over the ul, then the text for each tabs is revealed. Looks like: 
ul:hover li a .tabname { display: block; }

We can iterate over this if it needs  to be refined

File: app/js/components/systemMetadata/systemMetadata.jade [hunk 3]
```diff
@@ -1,11 +1,104 @@
...
+              b {{ data.label }}
+            card-content
+              ul.list.list-simple
+                li(ng-repeat="value in data.values", translate)
+                  strong {{value.label}}
+                  br
+                  span {{value.value}}
+
+
+
```
Comment (mattnolting) [inline app/js/components/systemMetadata/systemMetadata.jade]: Remove these lines after functionality is confirmed

File: app/js/components/systemMetadata/systemMetadata.jade [hunk 4]
```diff
@@ -1,11 +1,104 @@
...
+        //      th.ellipsis-overflow(ng-if="data.label === 'BIOS'", translate) BIOS
+        //      th.ellipsis-overflow(ng-if="data.label === 'System'", translate) System
+        //  tbody
+        //    tr
+        //      td(ng-repeat="value in data.values")
+        //        .dataName
+        //          span {{value.label}}
+        //        .dataValue
+        //          span {{value.value}}
+    #network.tab-pane(role='network')
```
Comment (mattnolting) [inline app/js/components/systemMetadata/systemMetadata.jade]: I couldn't see a working example of this, but what I saw looked a bit odd with the scrollbar outside of the table and the css looks overly complicated. I can see potential issues with choosing this solution.

A simpler and cleaner way to accomplish the same functionality, is to return a table of 3 rows with the option to show the rest if they exist. Upon click > the remaining rows are loaded

File: app/styles/components/system-overlay/_system-metadata.scss [hunk 5]
```diff
@@ -0,0 +1,29 @@
+.system-metadata {
+  //@include media-query(sm) { @include rem('padding-left', 25px); }
+
+  .metadata-value { word-wrap: break-word; }
+  .alert-danger { @include rem('margin-top', $global-padding); }
+
+  .basic-information { .icon-inline { @include rem('padding-left', 64px); } }
+  //.system-metadata { padding-bottom: $global-padding; }
```
Comment (mattnolting) [inline app/styles/components/system-overlay/_system-metadata.scss]: Remove commented code

File: app/styles/components/utilities/_expandable-data-table.scss [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    }
+  }
+
+  tbody {
+    @include rem('max-height', $tableSize);
+    overflow-y: scroll;
+  }
+}
+
+//.multipleSingleColumnTables,
```
Comment (mattnolting) [inline app/styles/components/utilities/_expandable-data-table.scss]: Not sure why custom scaffolding would be necessary here, bootstrap scaffolding works fine

**Comment by iphands** [comment]
@mattnolting Can we equal heights these cards?

@jeudy100 Still having issues where VMs that should be detected are not.

tmp

**Comment by iphands** [comment]
@jeudy100 the IP PORT stuff is broken here.

For a blank IP could you just say "Any" or "All"

In some cases the port looks like :1:4325 that is bad.
Note that IPv6 address short names can look like ::::::1234

tmp2

**Comment by mattnolting** [comment]
Should be ready to go. Made some changes for mobile fitting. @jeudy100 I noticed that if there are exactly three systems network table entries, the expand table link appears. Please double check.

**Comment by mattnolting** [comment]
@ryelo please check out branch and qa:)

**Comment by ryelo** [comment]
If you expand the tabs, select the networking tab, and then minimize it, the networking tab still shows as active. If you click the expand button again, it'll open up to the system tab. 

IMO we should remove the color of the tabs and make it look like the initial state if you minimize the tabs. 

Thoughts @mattnolting ?
screen shot 2017-09-25 at 9 13 45 am
screen shot 2017-09-25 at 9 13 23 am

**Comment by mattnolting** [comment]
@ryelo Agree, minimizing the tabs should remove all 'active' styling and return tabs to default state

**Comment by ryelo** [comment]
Everything looks good on my end!

**Comment by Kinlaw** [comment]
I haven't finished reviewing yet @jeudy100, @ryelo but two things so far:

1)  The jenkins integration is failing because of the merge conflict.
2)  The group tags positioning is overlapping the system box. I don't think it's worth holding the PR up for but if you can get a fix in, it would look better.

screen shot 2017-10-03 at 4 10 22 pm

**Comment by jeudy100** [comment]
@Kinlaw I am fixing all those issues. The server types are unknown because I had to make changes to expose more data collected by the engine/plugins for my npm module to work and that was merged today.

File: app/js/components/expandAllButton/expandAllButton.jade [hunk 7]
```diff
@@ -1,4 +1,4 @@
-a(ng-click="onClick()")
+exa(ng-click="onClick()")
```
[the remark this answers is not in the document]
Comment (iphands) [reply app/js/components/expandAllButton/expandAllButton.jade]: Looks like a typo right?

[the remark this answers is not in the document]
Comment (ryelo) [reply app/js/components/expandAllButton/expandAllButton.jade]: Looks like a typo

**Comment by ryelo** [comment]
@jeudy100 Have you been able to take a look at this on your end to make sure it looks okay?

File: app/js/components/expandAllButton/expandAllButton.jade (hunk 7 above)
[the remark this answers is not in the document]
Comment (Kinlaw) [reply app/js/components/expandAllButton/expandAllButton.jade]: can we change this to just a instead of exa?

**Comment by iphands** [comment]
Closing in favor of #275

**Pull request closed without merging**